### ISSUE TITLE
inferClassifier freezes

### ISSUE THREAD

**Issue by coolpaul** [markowetzlab/pathml#3]
Thank you very much for a great development!
The software installed fine and training executes well with gpu support.
However, following training, the slide inference step freezes....
Tried to reduce batch size, but this had no effect.
Presumably this is a pytorch issue?
Many thanks for your help.

start_time = time.time()

class_names = ['metastasis', 'non_metastasis']
means_and_stds = pickle.load(open(os.path.join(analysis_dir_path, 'classification_results', 'trainval_channel_means_and_stds.p'), 'rb'))
modelPath = glob.glob(os.path.join(analysis_dir_path, 'classification_results', 'classification_best_model*'))[0]

dataTransforms = transforms.Compose([
    transforms.Resize(224),
    transforms.ToTensor(),
    transforms.Normalize(means_and_stds['channel_means'], means_and_stds['channel_stds'])])

trainedModel = models.vgg19_bn(pretrained=False)
num_ftrs = trainedModel.classifier[6].in_features
trainedModel.classifier[6] = nn.Linear(num_ftrs, len(class_names))
trainedModel.load_state_dict(torch.load(modelPath))

for case in val_cases + test_cases:
    pathml_slide = Slide(os.path.join(analysis_dir_path, 'pathml_slides', case+'.pml'))
    ######## this step freezes......
    pathml_slide.inferClassifier(trainedModel, classNames=class_names, dataTransforms=dataTransforms, tissueLevelThreshold=0.995, foregroundLevelThreshold=88)
    pathml_slide.save(folder=pathml_slide_dir_path)
    
time_elapsed = time.time() - start_time
print('Complete in {:.0f}m {:.0f}s'.format(time_elapsed // 60, time_elapsed % 60))

OUTPUT
Inferring model on GPU
  0%|          | 0/60 [00:00<?, ?it/s]

SPEC
i9 with 16 CPU and 64GB RAM with GPU 
Ubuntu 20.04

GPU allocation:
'|===========================================================================|\n|                  PyTorch CUDA memory summary, device ID 0                 |\n|---------------------------------------------------------------------------|\n|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |\n|===========================================================================|\n|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |\n|---------------------------------------------------------------------------|\n| Allocated memory      |    2669 MB |    3418 MB |   11114 GB |   11111 GB |\n|       from large pool |    2663 MB |    3412 MB |   11086 GB |   11084 GB |\n|       from small pool |       5 MB |       6 MB |      27 GB |      27 GB |\n|---------------------------------------------------------------------------|\n| Active memory         |    2669 MB |    3418 MB |   11114 GB |   11111 GB |\n|       from large pool |    2663 MB |    3412 MB |   11086 GB |   11084 GB |\n|       from small pool |       5 MB |       6 MB |      27 GB |      27 GB |\n|---------------------------------------------------------------------------|\n| GPU reserved memory   |    2742 MB |    3544 MB |    4022 MB |    1280 MB |\n|       from large pool |    2736 MB |    3536 MB |    4012 MB |    1276 MB |\n|       from small pool |       6 MB |       8 MB |      10 MB |       4 MB |\n|---------------------------------------------------------------------------|\n| Non-releasable memory |   74728 KB |  218569 KB |    3400 GB |    3400 GB |\n|       from large pool |   74416 KB |  217124 KB |    3369 GB |    3369 GB |\n|       from small pool |     312 KB |    2924 KB |      31 GB |      31 GB |\n|---------------------------------------------------------------------------|\n| Allocations           |     446    |     501    |    1512 K  |    1511 K  |\n|       from large pool |      70    |     115    |     675 K  |     675 K  |\n|       from small pool |     376    |     430    |     837 K  |     836 K  |\n|---------------------------------------------------------------------------|\n| Active allocs         |     446    |     501    |    1512 K  |    1511 K  |\n|       from large pool |      70    |     115    |     675 K  |     675 K  |\n|       from small pool |     376    |     430    |     837 K  |     836 K  |\n|---------------------------------------------------------------------------|\n| GPU reserved segments |      37    |      61    |      65    |      28    |\n|       from large pool |      34    |      57    |      60    |      26    |\n|       from small pool |       3    |       4    |       5    |       2    |\n|---------------------------------------------------------------------------|\n| Non-releasable allocs |      18    |      48    |     767 K  |     767 K  |\n|       from large pool |      16    |      36    |     387 K  |     387 K  |\n|       from small pool |       2    |      26    |     380 K  |     380 K  |\n|---------------------------------------------------------------------------|\n| Oversize allocations  |       0    |       0    |       0    |       0    |\n|---------------------------------------------------------------------------|\n| Oversize GPU segments |       0    |       0    |       0    |       0    |\n|===========================================================================|\n'

**Comment by agberman** [comment]
Hi,

Thank you for the kind words, and for trying out PathML!

This is almost certainly a PyTorch bug. I've run into something just like this before, except not on this function, but the solution was to use only 1 worker. I have numWorkers is an argument in inferClassifier, try setting it to 1 and see if it works?

Best regards,
Adam

**Comment by coolpaul** [comment]
Thank you very much for your help. Unfortunately, setting numWorkers=0 or 1 or 2 doesn't help and it still freezes. Is there another option? Many thanks, 
Paul

**Comment by coolpaul** [comment]
Setting numWorkers=0 and reducing the batch size works!
Thanks very much for a great development!
Best wishes,
Paul

**Comment by agberman** [comment]
Hi Paul,

Ah, I see. Hmm. Let me poke around and test it some more myself. I will try to reproduce the freezing as it doesn't freeze for me at the moment, then I will report back. Thank you for your patience!

**Comment by agberman** [comment]
> Setting numWorkers=0 and reducing the batch size works!
> Thanks very much for a great development!
> Best wishes,
> Paul

Oh, I've just seen this! Glad to hear it is working for you now!
